### DISCUSSION TITLE
Input and TextArea have cursor blinking even though terminal window is not focused

### DISCUSSION THREAD

**Discussion by villekr** [Textualize/textual#4259] [Q&A]
Following simple app has Input and TextArea widgets. Focus transitions between them as expected. However, when I **focus some other window** than iTerm2 running textual app then cursor still keeps on blinking on the previously focused widget. I would expect that by default focus is removed alltogether from app when other window gets focus. Or is this something app needs to be specifically configured? 

<del>I added also _on_app_focus annd _on_app_blur event handlers for debugging, but they are not invoked. Maybe these are only relevant for textual-web based on comments (# Required by textual-web to manage focus in a web page.)</del>

As this might be platform/terminal related here's the configuration I have.

macOS Sonoma: Version 14.3
iTerm2: Build 3.4.23
Python 3.12
Textual 0.52.1

```
from textual import events
from textual.app import App, ComposeResult
from textual.containers import Vertical
from textual.widgets import Input, TextArea


class AppFocus(App):

    def compose(self) -> ComposeResult:
        with Vertical():
            yield TextArea()
            yield Input()


if __name__ == "__main__":
    AppFocus().run()

```

**Comment by davep** [comment]
The cursor within those controls isn't the terminal's own cursor; it is an effect provided by the widgets themselves. Those widgets can't know if the terminal emulator window you're using is focused in your environment.

To the best of my knowledge `AppFocus` and `AppBlur` are only supported when the application is being used via textual-web (this is how they're documented in the docs).

**Reply by villekr** [reply]
I understand. But the current behaviour is very confusing. It's not one or two times I've started to write text thinking that textual Input/TextArea widget has focus even though terminal window is not actually focused.

So what would be the approach to disable blinking effect when terminal window loses the focus? And also how to restore blinking effect when terminal window gets focus?

**Reply by davep** [reply]
I'm not personally aware of a method via which those widgets could know that the emulator doesn't have focus.

**Reply by villekr** [reply]
Sounds like a bug / missing feature to me.

**Reply by willmcgugan** [reply]
There is an ANSI sequence for this.(https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-FocusIn_FocusOut). Not sure how well supported it is.

**Reply by TomJGooding** [reply]
> There is an ANSI sequence for this.(https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-FocusIn_FocusOut). Not sure how well supported it is.

Interesting, it is a shame these features aren't better documented!

From a quick experiment with the terminals to hand (alacritty and urxvt), this seems to work. If it is widely supported this could be useful feature.

EDIT: Updated my experiment to disable the focus tracking again, which otherwise could be annoying!

```python
from textual.app import App, ComposeResult
from textual.widgets import TextArea


class TerminalFocusApp(App):
    def on_mount(self) -> None:
        # enable xterm focus tracking
        self._driver.write("\033[?1004h\n")  # type: ignore

    def compose(self) -> ComposeResult:
        # If the above xterm focus tracking is working, the TextArea should
        # show the unhandled FocusIn/FocusOut escape sequences after the
        # terminal window loses then gains focus
        yield TextArea()

    async def _on_exit_app(self) -> None:
        # disable xterm focus tracking again!
        self._driver.write("\033[?1004l\n")  # type: ignore


if __name__ == "__main__":
    app = TerminalFocusApp()
    app.run()
```

**Reply by davep** [reply]
On macOS:

- [x] Kitty[^1]
- [x] iTerm2
- [x] Contour
- [x] terminal.app
- [x] wezterm
- [x] rio
- [x] Tabby
- [x] VSCode terminal

On Windows 11 (testing the ones that don't overlap with the above):

- [x] Terminal
- [x] Hyper

While I suspect it was more down to how the underlying driver handles incoming stuff, I'd also note that these seemed to turn up at very unreliable times; without diving into the code right this moment I think that's likely down to how the driver approaches unknown escape sequences, but I thought it worth a note as something to be aware of (a "laggy" handling of app focus/blur would be worse than no handling).

[^1]: Also works with Kitty's own internal windowing facility, so splitting a window and swapping to another "pane" in the same focused OS-level window sent the relevant escape sequence.

**Reply by TomJGooding** [reply]
Surprisingly even the old Windows conhost seems to support it from a quick test.

**Reply by davep** [reply]
New challenge: find something that doesn't.

**Reply by TomJGooding** [reply]
Indeed...

Apparently iTerm has a "Apps may turn on Focus Reporting" option in the advanced settings -  might be worth trying that?

**Reply by davep** [reply]
Good catch.

**Comment by villekr** [comment]
Taken care in https://github.com/Textualize/textual/pull/4265. So this can be closed. Thanks for the textual devs for making this improvement!
